[FEATURES] ActivateOpenWindow, OpenWindowDetection, ChildLockEnabled

ilueckel opened the pull request:
This merge requests adds following features

# Child Lock
You can now enable/disable the child lock by setting the `childLockEnabled` property.

# Open Window Detection
You can now modify the settings `tado.[Z].[X].Rooms.[Y].openWindowDetection.enabled` and `tado.[Z].[X].Rooms.[Y].openWindowDetection.timeoutInSeconds`.

Note: I did not change state attribute `write` on `enabled` in the `state_attr.js`, as the parameter is used in other places as well. I guess, you can set it via code nonetheless?

# ActivateOpenWindow
When the thermostat detects an open window, tado will send you a notification and in the app appears a button to power off the radiator/thermostat (for the period defined in `tado.[Z].[X].Rooms.[Y].openWindowDetection.timeoutInSeconds`). The new button `tado.[Z].[X].Rooms.[Y].activateOpenWindow` can trigger this app call. 

Using the button you can create a blockly automation like this:
![image](https://user-images.githubusercontent.com/1216059/150647795-e9eb5cc1-b762-42da-9a5e-313cf10cbf23.png)
This basically is the same the "Auto-Assist" does.

Note that triggering `tado.[Z].[X].Rooms.[Y].activateOpenWindow` without a detected open window by tado has no effect.

HGlab01 commented:
Hi @ilueckel ! Thanks for the PR! I just recognized the PR today and will have a look on it in the next few days!

HGlab01 commented on main.js:
```diff
@@ -1056,7 +1135,7 @@ class Tado extends utils.Adapter {
 		if (method != 'get') {
 			this.apiCallinExecution = true;
-			console.log(`Body "${JSON.stringify(data)}" for API call "${url}"`);
+			this.log.debug(`Body "${JSON.stringify(data)}" for API call "${url}"`);
```
'console.log' was on purpose, this helps a lot for debugging sentry info. There the console log is available.
Logging both is an option.

ilueckel replied:
Makes sense. I used the iobroker logs to debug everything, therefore I changed it

HGlab01 commented on main.js:
```diff
@@ -123,6 +128,8 @@ class Tado extends utils.Adapter {
 						const power = await this.getStateAsync(home_id + '.Rooms.' + zone_id + '.setting.power');
 						const durationInSeconds = await this.getStateAsync(home_id + '.Rooms.' + zone_id + '.overlay.termination.durationInSeconds');
 						const nextTimeBlockStart = await this.getStateAsync(home_id + '.Rooms.' + zone_id + '.nextTimeBlock.start');
+						const openWindowDetectionEnabled = await this.getStateAsync(home_id + '.Rooms.' + zone_id + '.openWindowDetection.enabled');
```
is there a reason not to put this in a separate "else if" block as  for 'childLockEnabled'?
In the last block all states are handled which trigger setZoneOverlay().
I suggest to add additional 'else if' blocks above...

ilueckel replied:
Okay, I wasn't sure about that. I added two new if-blocks for `activateOpenWindow` and `openWindowDetection`-settings

HGlab01 commented on README.md:
```diff
@@ -20,22 +20,26 @@ If you like my work, please feel free to provide a personal donation
-| tado.[x].[yyyyyy].Rooms.[z].setting.horizontalSwing | Horizontal swing (only AC devices with V3 and olderversions) |
+| State                                                                   | Description                                                                                                  |
+|-------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------|
+| tado.[x].[yyyyyy].Rooms.[z].setting.power                               | Turn device on/off                                                                                           |
```
Which advantage do you see in adding the whites paces?
On my laptop screen one row is now displayed in 2 lines, which makes it harder to read

ilueckel replied:
Wasn't done on purpose, IntelliJ seems to update this automatically. As long as you don't wrap lines in the IDE, it increases readability as all columns are aligned. I changed it back

ilueckel commented:
Hi @HGlab01,
I was wondering if there is anything left or needs to be changed in handling the states from your point of view?

HGlab01 commented:
Really nice implementations!
I hope it was not to hard to understand the general approaches for your additional implementations!

@state enabled:
I did the following "trick" for the attribute celsius:
https://github.com/DrozmotiX/ioBroker.tado/blob/dde43947766ba063d7905322ef699ae836e74e6e/main.js#L821-L824
With this I renamed the JSON attribute before creating states. I think you can do this here as well.
Maybe not extremely beautiful but it works ;-)

ilueckel commented:
Great - yeah this would be a good possibility. I renamed `openWindowDetection.enabled` to `openWindowDetection.openWindowDetectionEnabled` and added the matching state.

HGlab01 commented:
Thanks for all the work! Let's merge to main and create an alpha version including your PR!

HGlab01 commented:
One additional remark/wish:
If you add additional things please deactivate sentry in your test installation. Otherwise I have a lot of issues in Sentry which makes me nervous until I recognize where it comes from ;-)

HGlab01 commented:
Enjoy!
https://github.com/DrozmotiX/ioBroker.tado/archive/refs/tags/v0.3.15-alpha.2.tar.gz
